elastic/kibana
# [Dashboard] Fix Embeddable Container Race Condition

### Summary
https://github.com/elastic/kibana/pull/126649 introduced a race condition where the initial creation of child embeddables could be interrupted when the container's input was updated during initialization. This would result in _multiple_ child embeddable being created, with only one of the resulting embeddables being rendered.

After this PR, child embeddables should be created _once_ and _only once_ no matter how many times the input is changed, and no matter how long it takes for the initial child creation to be complete. Additionally, if one or more embeddables are added _during_ the initialization phase, they will be properly added after initialization is completed.

To accomplish this, this PR:

1. Changes the `initializeChildEmbeddables` function to subscribe to any input changes, but wait for initialization of all child embeddables to be **complete** before checking if its children should update.
2. The `initializeChildEmbeddables` function will rely on the subscription to initialize children if there is no requirement for children to be initialized in order.
3. Changes the `ControlGroupContainer` to use a function that memoizes the order value instead of a member variable cache on the class. This is so that the order cache can be used synchronously from the constructor. This speeds up embeddable container loading slightly, because child initialization no longer needs to be delayed.

This PR also adds a new embeddable container jest test, and updates the rest of them to spy on the factory's create method rather than the `onPanelAdded` method. This was done so that any embeddable creations which happen in the embeddable container's constructor can be properly captured.

### How to test
To test that every embeddable is created **once and only once** you can add a console log to the `onPanelAdded` method in `src/plugins/embeddable/public/lib/containers/container.ts` which logs the ID / title of the panel. When loading a dashboard with a control group, you should see the same number of logs as there are children (including dashboard panels, and Controls) and you should **not** see duplicate ids.

### Checklist

Delete any items that are not applicable to this PR.
